### ISSUE TITLE
Custom Post Type with number breaks Schema

### ISSUE THREAD

**Issue by rubas** [wp-graphql/wp-graphql#1346]
If you register a custom post type with a name like `111-222` (stupid, but legitimate), it will break the graphql schema as the name is interpreted as int.

graphql-codegen output:
```
Failed to load schema from https://local.test/wp/graphql:

Syntax Error: Expected Name, found Int "111222".

GraphQLError: Syntax Error: Expected Name, found Int "111222".
at syntaxError (node_modules/graphql/error/syntaxError.js:15:10)
```

---

Setting `graphql_single_name` and `graphql_plural_name` doesn't change this.

**Comment by jasonbahl** [comment]
@rubas thanks for opening this issue! This is indeed a case I had not considered. 

Can you provide the exact code you were using to register the post types so I can replicate?

I'm able to get this to work using `graphql_single_name` and `graphql_plural_name`

Below you can see how it shows in various ways in the Schema: 

![Screen Shot 2020-06-25 at 8 04 37 AM](https://user-images.githubusercontent.com/1260765/85734859-92208b00-b6ba-11ea-888e-bf5a16abad90.png)

**Comment by rubas** [comment]
```
register_post_type('1995-2035', [
            'labels' => [
              'name'          => '1995-2035',
              'singular_name' => '1995-2035',
              'menu_name'     => '1995-2035',
            ],
            'public'              => true,
            'map_meta_cap'        => true,
            'query_var'           => true,
            'show_ui'             => true,
            'show_in_rest'        => false,
            'menu_icon'           => 'dashicons-backup',
            'menu_position'       => 10,
            'show_in_graphql'     => false,
            'graphql_single_name' => 'CH1995',
            'graphql_plural_name' => 'CH1995',
            'supports'            => [
              0 => 'title',
              1 => 'editor',
              2 => 'author',
              3 => 'revisions',
            ],
            'hierarchical' => false,
            'taxonomies'   => [
            ],
            'has_archive' => false,
            'can_export'  => true,
          ]);
```

**Comment by rubas** [comment]
Changing this line fixes the "template" naming problem: 

https://github.com/wp-graphql/wp-graphql/blob/develop/src/Type/Union/ContentTemplateUnion.php#L25

**Comment by jasonbahl** [comment]
Ah. So it's a template you were using? Can you share the template so I can replicate that part to? or at least the top part of the template that declares the template name, etc?

**Comment by rubas** [comment]
... Ah, now we are talking. We have multiple issues at the same time. Sorry, got confused.

a) Post Type named `1995-2035`
b) Template named `1995-2035`

To fix both issues, I have disable the post_type with `show_in_graphql: false` and change the name of the template.

**Comment by jasonbahl** [comment]
Thanks for the info. I'll try to replicate right now.

**Comment by jasonbahl** [comment]
@rubas This is an interesting situation. I'm not sure I have the right answers yet. 

We do need to ensure that GraphQL Field and Type names start with a letter, per the GraphQL Spec: http://spec.graphql.org/draft/#Letter

I'm not sure what the best way to enforce this is. 🤔 

I'll need to think about it a bit. 

For post types and taxonomies, the `graphql_single_name` and `graphql_plural_name` fields help because you can at least change the names for the GraphQL Schema while leaving the rest of the post type to behave as it was for other parts of WordPress. 

For templates, right now since we generate the name from the template name, you had to change the template name, and that can have side-effects that could be frustrating. Folks shouldn't need to change template names to use WPGraphQL. 

Perhaps we can detect a `GraphQL Name` in the template. 

Something like: 

```php
<?php
/**
 * Template Name: 1995-2035
 * GraphQL Name: Template1995to2035
 */
```

Or something like that? 🤔 

The other thought is that we change the order of `$templateName . Template` to `Template . $templateName`

That way the Schema would have: 

`TemplateDefault`, `Template1995-2035`, etc.

**Comment by jasonbahl** [comment]
@ rubas I also noticed you're using the same name for `graphql_single_name` and `graphql_plural_name`: 

```
'graphql_single_name' => 'CH1995',
'graphql_plural_name' => 'CH1995',
```
This is also is problematic because there are singular entry points added to the graph using the single name, and plural connection entry points added using the plural name. 

This should probably also be better documented, and also throw some sort of exception so users know to update their Schema.

**Comment by rubas** [comment]
> For post types and taxonomies, the graphql_single_name and graphql_plural_name fields help because you can at least change the names for the GraphQL Schema while leaving the rest of the post type to behave as it was for other parts of WordPress.

I does not. Without `show_in_graphql` set `true`. I get the following error

```
Syntax Error: Expected Name, found Int "19952035" (343:3) 341 | """The CH19952035 type""" 342 | type CH19952035 implements Node & ContentNode & UniformResourceIdentifiable & NodeWithTitle & NodeWithContentEditor & NodeWithAuthor & NodeWithRevisions { > 343 | 19952035: CH19952035_19952035 | ^ 344 | 345 | """The id field matches the WP_Post->ID field.""" 346 | cH19952035Id: Int! @deprecated(reason: "Deprecated in favor of the databaseId field")
```
---

> The other thought is that we change the order of $templateName . Template to Template . $templateName

I definitely prefer this over parsing the PHP template files (read php files, tokenize, analyse it ..😕)

**Comment by jasonbahl** [comment]
Ok, @TylerBarnes and I have been discussing this a bit. 

Our current thought is that if a Type or Field is invalid (starts with an integer, for example) we will prevent it from being added to the Schema and will add the fields to some sort of debug message. 

So, in a case like this where `111222` is invalid, the Schema will generate without it, the Field/Type will be logged in some debug log that will be output with requests, but _not_ as errors. 

Some response like: 

```
{
  "data": ...
  "errors": ...
  "debug": [
    {
      message: "Field 111222 on Type RootQuery was not added to the Schema because fields cannot start with an integer",
    }
  ]
}
```

This will allow production applications to still execute, make use of data/errors as needed, but the debug info can be used to fix underlying issues, such as fields that were not added to the Schema because of invalid field/type names, etc.

**Comment by jasonbahl** [comment]
closed by #1555
